Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs

```diff
@@ -0,0 +1,115 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Content.Server.Database;
+
+namespace Content.Server.Administration.Managers;
+
+public sealed partial class UsernameRuleManager
+{
+    public const string UsernameRuleNotificationChannel = "username_rule_notification";
+
+    private static readonly TimeSpan UsernameRuleNotificationRateLimitTime = TimeSpan.FromSeconds(30);
+    private const int UsernameRuleNotificationRateLimitCount = 10;
+
+    private readonly object _usernameRuleNotificationRateLimitStateLock = new();
+    private TimeSpan _usernameRuleNotificationRateLimitStart;
+    private int _usernameRuleNotificationRateLimitCount;
+
+    private void OnDatabaseNotification(DatabaseNotification notification)
+    {
+        if (notification.Channel != UsernameRuleNotificationChannel)
+        {
+            return;
+        }
+
+        if (notification.Payload == null)
+        {
+            _sawmill.Error("got username rule notification with no payload");
+            return;
+        }
+
+        UsernameRuleNotification data;
+        try
+        {
+            data = JsonSerializer.Deserialize<UsernameRuleNotification>(notification.Payload)
+                ?? throw new JsonException("Content is null");
+        }
+        catch (JsonException e)
+        {
+            _sawmill.Error($"Got invalid JSON in username rule notification: {e}");
+            return;
+        }
+
+        if (!CheckRuleRateLimit())
+        {
+            _sawmill.Verbose("Not processing username rule notification due to rate limit");
+            return;
+        }
+
+        _taskManager.RunOnMainThread(() => ProcessUsernameRuleNotification(data));
+    }
+
+    private async void ProcessUsernameRuleNotification(UsernameRuleNotification data)
+    {
+        if ((await _entryManager.ServerEntity).Id == data.ServerId)
+        {
+            _sawmill.Verbose("Not processing username rule notification: came from this server");
+            return;
+        }
+
+        _sawmill.Verbose($"Processing username rule notification for username rule {data.UsernameRuleId}");
+        var usernameRule = await _db.GetServerUsernameRuleAsync(data.UsernameRuleId);
+        if (usernameRule == null)
+        {
+            _sawmill.Warning($"Username rule in notification ({data.UsernameRuleId}) didn't exist?");
+            return;
+        }
+
+        if (usernameRule.Retired)
+        {
+            ClearCompiledRegex(usernameRule.Id ?? -1);
+            return;
+        }
+
+        CacheCompiledRegex(usernameRule.Id ?? -1, usernameRule.Regex, usernameRule.Expression, usernameRule.Message, usernameRule.ExtendToBan);
```

Comment:
Isn't this going to log a warning when updating an existing rule with a new regex? Is that a supported use case?